src/main.cpp

```diff
@@ -5930,32 +5930,38 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
             pfrom->fDisconnect = true;
             return true;
         }
+      
         std::vector<uint256> vtxid;
-        if (1)  // Keep this lock for as short as possible, causing 2.6 second locks
-	  {
-          LOCK2(cs_main, pfrom->cs_filter);
-          mempool.queryHashes(vtxid);
-	  }
         vector<CInv> vInv;
-        BOOST_FOREACH(uint256& hash, vtxid) {
-            CInv inv(MSG_TX, hash);
-            if (pfrom->pfilter) {
-                CTransaction tx;
-                bool fInMemPool = mempool.lookup(hash, tx);
-                if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
-                if (!pfrom->pfilter->IsRelevantAndUpdate(tx)) continue;
-            }
-            vInv.push_back(inv);
-            if (vInv.size() == MAX_INV_SZ) {
-                pfrom->PushMessage(NetMsgType::INV, vInv);
-                vInv.clear();
+        mempool.queryHashes(vtxid); // BU: No lock required here. This has it's own internal lock
+        {
+            LOCK(pfrom->cs_filter);
+            BOOST_FOREACH(uint256& hash, vtxid) {
+                CInv inv(MSG_TX, hash);
+                if (pfrom->pfilter) {
+                    CTransaction tx;
+                    // BU: the following is commented out and is not worth the effort.  If we do this will
+                    //     repeatedly lock and unlock the memory pool.
+                    //bool fInMemPool = mempool.lookup(hash, tx);
+                    //if (!fInMemPool) continue; // another thread removed since queryHashes, maybe...
+                    if (!pfrom->pfilter->IsRelevantAndUpdate(tx)) continue;
```

Comment:
mempool.lookup fills tx with the transaction, so it cannot be commented out because this line uses the tx.

To reduce inner-loop locking its a pretty common convention to preface lockless versions of functions with _, so I'd recommend that you make a mempool::_lookup() function and call it instead of .lookup()